# Pandas Introduction

![image](../../images/pandas_logo.png)

**Pandas** is probably the most popular Python library for data manipulation and analysis. Pandas is designed to work with a wide range of data sources, including: `CSV`, `Excel`, `Parquet`, `Pickle`, `JSON`, and many more.

**NOTE**: Data Scientists / Data Professionals 💚 **Pandas**!

# Install and Use Pandas for the first time

`pandas` should be automatically installed as part of `anaconda`. Nonetheless, if for some reasons it is missing, you can enter the following command in a `anaconda prompt` or a `terminal` to install `pandas`:

```bash
pip install pandas
```

You can verify that it is installed by entering the following codes in a cell in a notebook:

```python
try:
    import pandas as pd
    print(f"The version of pandas is: {pandas.__version__}")
except ImportError:
    print("pandas is not installed!")
```

In [1]:
try:
    import pandas as pd
    print(f"The version of pandas is: {pd.__version__}")
except ImportError:
    print("pandas is not installed!")

The version of pandas is: 1.4.2


It is extremely common to use `pd` as the alias for `pandas` in `Python`.

```python
import pandas as pd
```

# Core Objects of Pandas

There are two **core objects** of `pandas`:
- **DataFrame**: A table-like object that can be easily created and manipulated.
- **Series**: A 1-dimensional sequence of data values. If a **dataframe** is a table, then each column is a **Series**.

![image](../../images/df_series.png)

## DataFrame

In order to **create a dataframe**, you can use `pd.DataFrame()` constructor.

Let's create the dataframe we see in the image above and assign it to a variable called `df`.

```python
df = pd.DataFrame({
    "Height": [1.60, 1.76, 1.95],
    "Weight": [60.0, 56.0, 85.0],
})

df
```

As you can see, we provide a dictionary of **column names** and **column values** to the `pd.DataFrame()` constructor. The **column names** are the **keys** of the dictionary, and the **column values** are the **values** of the dictionary. This is a very common way to create a dataframe.

In [2]:
df = pd.DataFrame({
    "Height": [1.60, 1.76, 1.95],
    "Weight": [60.0, 56.0, 85.0],
})

df

Height  Weight
0    1.60    60.0
1    1.76    56.0
2    1.95    85.0

The numbers in the far left of the dataframe are indices. The indices are automatically assigned by `pandas` and are called **row labels**.

In this example, the indices are `0`, `1`, and `2`. The [`1`, `Weight`] entry contains the value `56.0`.

**NOTE**:
- Dataframes can contain as many columns and rows as you want. The entries in the dataframes are not limited to numbers, and they can be of any type.
- Dataframes are **mutable**. This means that you can change the values of the dataframe.

**Exercise**

Given the following dictionary, can you create a dataframe that looks like the one below?

```python
dict_data = {
    "Name": ["Millie Bobby Brown", "Finn Wolfhard", "Natalie Dyer", "Kaley Cuoco", "Jim Parsons"],
    "Age": [18, 19, 27, 36, 49],
    "Height": [1.60, 1.80, 1.63, 1.68, 1.86],
    "TV Series": ["Stranger Things", "Stranger Things", "Stranger Things", "Big Bang Theory", "Big Bang Theory"],
}
```

|Name|Age|Height|TV_Series|
|:-:|:-:|:-:|:-:|
|Millie Bobby Brown|18|1.60|Stranger Things|
|Finn Wolfhard|19|1.80|Stranger Things|
|Natalie Dyer|27|1.63|Stranger Things|
|Kaley Cuoco|36|1.68|Big Bang Theory|
|Jim Parsons|49|1.86|Big Bang Theory|

In [3]:
# [TODO]
df = pd.DataFrame({
    "Name": ["Millie Bobby Brown", "Finn Wolfhard", "Natalie Dyer", "Kaley Cuoco", "Jim Parsons"],
    "Age": [18, 19, 27, 36, 49],
    "Height": [1.60, 1.80, 1.63, 1.68, 1.86],
    "TV Series": ["Stranger Things", "Stranger Things", "Stranger Things", "Big Bang Theory", "Big Bang Theory"],
})

df

Name  Age  Height        TV Series
0  Millie Bobby Brown   18    1.60  Stranger Things
1       Finn Wolfhard   19    1.80  Stranger Things
2        Natalie Dyer   27    1.63  Stranger Things
3         Kaley Cuoco   36    1.68  Big Bang Theory
4         Jim Parsons   49    1.86  Big Bang Theory

Instead of using generic numbers as **row labels**, can we create a dataframe having row labels to be the names of the people in the dataframe?

The answer is **YES**, and it's very easy.

```python
df = pd.DataFrame({
    "Age": [18, 19, 27, 36, 49],
    "Height": [1.60, 1.80, 1.63, 1.68, 1.86],
    "TV Series": ["Stranger Things", "Stranger Things", "Stranger Things", "Big Bang Theory", "Big Bang Theory"],
}, index=["Millie Bobby Brown", "Finn Wolfhard", "Natalie Dyer", "Kaley Cuoco", "Jim Parsons"])

df
```

In [4]:
df = pd.DataFrame({
    "Age": [18, 19, 27, 36, 49],
    "Height": [1.60, 1.80, 1.63, 1.68, 1.86],
    "TV Series": ["Stranger Things", "Stranger Things", "Stranger Things", "Big Bang Theory", "Big Bang Theory"],
}, index=["Millie Bobby Brown", "Finn Wolfhard", "Natalie Dyer", "Kaley Cuoco", "Jim Parsons"])

df

Age  Height        TV Series
Millie Bobby Brown   18    1.60  Stranger Things
Finn Wolfhard        19    1.80  Stranger Things
Natalie Dyer         27    1.63  Stranger Things
Kaley Cuoco          36    1.68  Big Bang Theory
Jim Parsons          49    1.86  Big Bang Theory

As you can see, we only have to provide the list of names as values of the `index` argument in the `pd.DataFrame()` constructor.

## Series

In order to **create a series**, you can use `pd.Series()` constructor.

```python
height_series = pd.Series([1.60, 1.80, 1.63, 1.68, 1.86])
```

As mentioned above, a series can be considered as a column in a dataframe. 
- A series can contain as many entries as you want.
- A series also has `index` that are automatically generated by `pandas` but can be changed as you wish. 
- A series can be of any type.
- A series has one overall `name` that is used to identify the series.

In [5]:
height_series = pd.Series([1.60, 1.80, 1.63, 1.68, 1.86])
height_series

0    1.60
1    1.80
2    1.63
3    1.68
4    1.86
dtype: float64

Let's recreate height_series, but we'll give it a `name` of `Height` this time.

```python
height_series = pd.Series([1.60, 1.80, 1.63, 1.68, 1.86], name="Height")
```

In [6]:
# let's recreate height_series but we'll give it a name this time
height_series = pd.Series([1.60, 1.80, 1.63, 1.68, 1.86], name="Height")
height_series

0    1.60
1    1.80
2    1.63
3    1.68
4    1.86
Name: Height, dtype: float64

# Read data from file

It's good that you know how to create a dataframe/series from scratch. In real life, you will probably be doing that only when you want to create some very small tests specific to your need. In this section, we'll learn how to use `pandas` to read data from a **CSV file** and store the data in a dataframe.

**CSV file** is a common file format used to store data. It is a simple text file that contains a list of comma separated values.

In the `data` folder, I have already provided a CSV file called `weatherAUS.csv`. We'll be using this data for the rest of the lesson.

![image](../../images/data_tree.png)

The **Weather in Australia** dataset is a public dataset. More information about the **Weather in Australia** dataset is available here [here](https://www.kaggle.com/jsphyg/weather-dataset-rattle-package).

**NOTE**: **Kaggle** is probably the best source of learning for all **data enthusiasts**.

The data contains daily weather observations taken from various weather stations in Australia from `1st November 2007` to `25th June 2017`.

The data includes `145k` rows of weather observation data. Each row represents a unique observation set and each observation set contains the attributes shown in the table below

|No|Attribute|Description|Type|
|-|-|-|-|
|1|Date|The date of observation|Data time|
|2|Location|The common name of the location of the weather station|String|
|3|MinTemp|Minimun temperature in $^\circ C$|Float|
|4|MaxTemp|Maximum temperature in $^\circ C$|Float|
|5|Rainfall|Amount of rainfall for the day in $mm$|Float|
|6|Evaporation|class A pan evaporation in $mm$ in the 24 hours to 9am|Float|
|7|Sunshine|Number of hours of bright sunshine in the day|Float|
|8|WindGustDir|Direction of the strongest wind gust in the 24 hours to midnight|String|
|9|WindGustSpeed|Speed in $Km/hr$ of the strongest wind gust in the 24 hours to midnight|Float|
|10|WindDir9am|Direction of the wind at 9am|String|
|11|WindDir3pm|Direction of the wind at 3pm|String|
|12|WindSpeed9am|Wind speed in $km/hr$ averaged over 10 minutes prior to 9am|Float|
|13|WindSpeed3pm|Wind speed in $km/hr$ averaged over 10 minutes prior to 3pm|Float|
|14|Humidity9am|Humidity in percentage at 9am|Float|
|15|Humidity3pm|Humidity in percentage at 3pm|Float|
|16|Pressure9am|Atmospheric pressure in $hpa$ reduced to mean sea level at 9am|Float|
|17|Pressure3pm|Atmospheric pressure in $hpa$ reduced to mean sea level at 3pm|Float|
|18|Cloud9am|Fraction of sky obscured by cloud at 9am, measured in $oktas$ in a range 0 (sky clear) to 8 (sky completely overcast)|Float|
|19|Cloud3pm|Fraction of sky obscured by cloud at 3pm, measured in $oktas$ in a range 0 (sky clear) to 8 (sky completely overcast)|Float|
|20|Temp9am|Temperature in $^\circ C$ at 9am|Float|
|21|Temp3pm|Temperature in $^\circ C$ at 3pm|Float|
|22|RainToday|Boolean attribute. 'Yes' if precipitation (mm) in the 24 hours to 9am exceeds 1mm, otherwise 'No'|String|
|23|RainTomorrow|The target variable. Did it rain tomorrow? Boolean. 'Yes' if we predict rain for tomorrow, otherwise 'No'|String|

Let's read the data from the CSV file and play with the data.

```python
weather_data = pd.read_csv("../../data/weatherAUS.csv")
```

In [7]:
weather_data = pd.read_csv("../../data/weatherAUS.csv", parse_dates=["Date"])

Similar to `np.ndarray`, `pd.DataFrame` also has `shape` attribute that can be used to check the number of rows and columns in the dataframe.

```python
weather_data.shape
```

In [8]:
weather_data.shape

(145460, 23)

Thus, we can see that the dataframe has `145,460` rows and `23` columns.

In order to examine the data, we can use the `head()`, `tail()`, or `sample()` method.
- `head(n)` returns the first `n` rows of the dataframe.
- `tail(n)` returns the last `n` rows of the dataframe.
- `sample(n)` returns `n` random rows of the dataframe.

Let's view the first `5` rows of the `weather_data` dataframe.

```python
weather_data.head()
```

In [9]:
weather_data.head()

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0 2008-12-01   Albury     13.4     22.9       0.6          NaN       NaN   
1 2008-12-02   Albury      7.4     25.1       0.0          NaN       NaN   
2 2008-12-03   Albury     12.9     25.7       0.0          NaN       NaN   
3 2008-12-04   Albury      9.2     28.0       0.0          NaN       NaN   
4 2008-12-05   Albury     17.5     32.3       1.0          NaN       NaN   

  WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  Humidity3pm  \
0           W           44.0          W  ...        71.0         22.0   
1         WNW           44.0        NNW  ...        44.0         25.0   
2         WSW           46.0          W  ...        38.0         30.0   
3          NE           24.0         SE  ...        45.0         16.0   
4           W           41.0        ENE  ...        82.0         33.0   

   Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  \
0       1007.7       1007.1       8.0       NaN     16.9     21.8         No   
1       1010.6       1007.8       NaN       NaN     17.2     24.3         No   
2       1007.6       1008.7       NaN       2.0     21.0     23.2         No   
3       1017.6       1012.8       NaN       NaN     18.1     26.5         No   
4       1010.8       1006.0       7.0       8.0     17.8     29.7         No   

   RainTomorrow  
0            No  
1            No  
2            No  
3            No  
4            No  

[5 rows x 23 columns]

**Quick observation**:
- You can see that there are lots of values `NaN` in the dataframe. `NaN` stands for `Not a Number`, and it means that the value is not available. We'll learn how to deal with `NaN` values later.
- The dataframe is also **too long** to be displayed in a single screen. That's why we see a `...` section in the middle of the dataframe. 

**Exercise**

Can you get 10 random rows from the `weather_data` dataframe?

In [10]:
# [TODO]
weather_data.sample(10)

Date      Location  MinTemp  MaxTemp  Rainfall  Evaporation  \
85744  2013-07-01      Brisbane     13.1     17.9       8.8          0.8   
136339 2009-08-26  AliceSprings     12.4     19.7       3.6          9.8   
77202  2015-05-01      Portland      6.2     22.0       0.2          3.2   
61102  2017-04-07       Bendigo      8.4     26.2       0.0          NaN   
59713  2013-06-18       Bendigo      1.7     14.3       0.0          1.1   
43341  2011-02-03    Wollongong     23.4     35.2       0.0          NaN   
81278  2009-10-08      Dartmoor      5.9     13.3       0.6          2.4   
62784  2013-08-19          Sale      7.3     13.2       1.2          7.2   
91665  2012-11-25     GoldCoast     20.4     27.6       0.0          NaN   
123149 2015-08-14         Perth      6.1     23.1       0.0          1.2   

        Sunshine WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  \
85744        0.8         WSW           28.0         SW  ...        93.0   
136339       0.0           E           35.0        WNW  ...        85.0   
77202        8.3         NNE           33.0          N  ...        79.0   
61102        NaN           N           50.0        ENE  ...        56.0   
59713        NaN         SSE           39.0        SSE  ...        88.0   
43341        NaN           S           59.0         NW  ...        91.0   
81278        5.7         SSE           37.0        SSE  ...        69.0   
62784        5.5         WNW           69.0        WNW  ...        62.0   
91665        NaN          NE           39.0         NE  ...        52.0   
123149       9.8         SSW           17.0        ENE  ...        82.0   

        Humidity3pm  Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  \
85744          90.0       1014.1       1010.6       7.0       8.0     14.3   
136339         87.0       1022.2       1018.3       8.0       8.0     12.6   
77202          50.0       1025.3       1020.8       NaN       NaN     10.9   
61102          31.0       1025.5       1021.1       8.0       8.0     15.9   
59713          61.0       1017.9       1018.0       1.0       1.0      8.4   
43341          51.0       1011.3       1009.7       8.0       NaN     23.9   
81278          68.0       1030.5       1030.1       NaN       NaN      9.8   
62784          50.0       1007.4       1007.0       1.0       5.0     10.8   
91665          56.0       1019.1       1016.6       NaN       NaN     26.3   
123149         45.0       1025.1       1020.9       0.0       1.0     12.7   

        Temp3pm  RainToday  RainTomorrow  
85744      16.4        Yes           Yes  
136339     13.7        Yes           Yes  
77202      21.5         No            No  
61102      24.7         No            No  
59713      13.2         No            No  
43341      34.4         No            No  
81278      11.8         No            No  
62784      10.6        Yes           Yes  
91665      26.1         No            No  
123149     22.4         No            No  

[10 rows x 23 columns]

Similar to `np.ndarray`, `pd.DataFrame` can also be tranposed. This is useful when you want to display the dataframe in a more readable format.

Let's view the transposed version of the first 5 rows.

```python
weather_data.head().T
```

In [11]:
weather_data.head().T

0                    1                    2  \
Date           2008-12-01 00:00:00  2008-12-02 00:00:00  2008-12-03 00:00:00   
Location                    Albury               Albury               Albury   
MinTemp                       13.4                  7.4                 12.9   
MaxTemp                       22.9                 25.1                 25.7   
Rainfall                       0.6                  0.0                  0.0   
Evaporation                    NaN                  NaN                  NaN   
Sunshine                       NaN                  NaN                  NaN   
WindGustDir                      W                  WNW                  WSW   
WindGustSpeed                 44.0                 44.0                 46.0   
WindDir9am                       W                  NNW                    W   
WindDir3pm                     WNW                  WSW                  WSW   
WindSpeed9am                  20.0                  4.0                 19.0   
WindSpeed3pm                  24.0                 22.0                 26.0   
Humidity9am                   71.0                 44.0                 38.0   
Humidity3pm                   22.0                 25.0                 30.0   
Pressure9am                 1007.7               1010.6               1007.6   
Pressure3pm                 1007.1               1007.8               1008.7   
Cloud9am                       8.0                  NaN                  NaN   
Cloud3pm                       NaN                  NaN                  2.0   
Temp9am                       16.9                 17.2                 21.0   
Temp3pm                       21.8                 24.3                 23.2   
RainToday                       No                   No                   No   
RainTomorrow                    No                   No                   No   

                                 3                    4  
Date           2008-12-04 00:00:00  2008-12-05 00:00:00  
Location                    Albury               Albury  
MinTemp                        9.2                 17.5  
MaxTemp                       28.0                 32.3  
Rainfall                       0.0                  1.0  
Evaporation                    NaN                  NaN  
Sunshine                       NaN                  NaN  
WindGustDir                     NE                    W  
WindGustSpeed                 24.0                 41.0  
WindDir9am                      SE                  ENE  
WindDir3pm                       E                   NW  
WindSpeed9am                  11.0                  7.0  
WindSpeed3pm                   9.0                 20.0  
Humidity9am                   45.0                 82.0  
Humidity3pm                   16.0                 33.0  
Pressure9am                 1017.6               1010.8  
Pressure3pm                 1012.8               1006.0  
Cloud9am                       NaN                  7.0  
Cloud3pm                       NaN                  8.0  
Temp9am                       18.1                 17.8  
Temp3pm                       26.5                 29.7  
RainToday                       No                   No  
RainTomorrow                    No                   No

# Indexing, Selecting, and Assigning in Pandas

## Indexing

There are **several ways** to access a column in the dataframe.
- `.` operator: `df.column_name` returns a series with the values of the column `column_name`.
- `[]` (indexing) operator: `df["column_name"]` does the same thing as the above.

Let's access the `Location` column of the `weather_data` dataframe.

```python
weather_data.Location
weather_data["Location"]
```

In [12]:
weather_data.Location

0         Albury
1         Albury
2         Albury
3         Albury
4         Albury
           ...  
145455     Uluru
145456     Uluru
145457     Uluru
145458     Uluru
145459     Uluru
Name: Location, Length: 145460, dtype: object

In [13]:
weather_data["Location"]

0         Albury
1         Albury
2         Albury
3         Albury
4         Albury
           ...  
145455     Uluru
145456     Uluru
145457     Uluru
145458     Uluru
145459     Uluru
Name: Location, Length: 145460, dtype: object

Personally, I prefer to use the indexing syntax (`df["column_name"]`) because it is more readable and extremely clear that we are accessing a column from the dataframe. In addition, the indexing syntax can handle cases where the column name contains reserved characters like `spaces`.

For example, `df["column name"]` will work, but `df.column name` will not.

`pandas` has built-in accessor operators called `iloc` and `loc` for more advanced operations.
- `iloc` is **position-based** selection, while `loc` is **label-based** selection.
- `iloc` is short for `integer location` and is used to select rows and columns by their index (position).
- Both `iloc` and `loc` are **row-first, column-second**. 

Let's access the first row of the `weather_data` dataframe using `iloc`.

```python
weather_data.iloc[0]
```

In [14]:
weather_data.iloc[0]

Date             2008-12-01 00:00:00
Location                      Albury
MinTemp                         13.4
MaxTemp                         22.9
Rainfall                         0.6
Evaporation                      NaN
Sunshine                         NaN
WindGustDir                        W
WindGustSpeed                   44.0
WindDir9am                         W
WindDir3pm                       WNW
WindSpeed9am                    20.0
WindSpeed3pm                    24.0
Humidity9am                     71.0
Humidity3pm                     22.0
Pressure9am                   1007.7
Pressure3pm                   1007.1
Cloud9am                         8.0
Cloud3pm                         NaN
Temp9am                         16.9
Temp3pm                         21.8
RainToday                         No
RainTomorrow                      No
Name: 0, dtype: object

Let's access the `MaxTemp` column of `weather_data` using the `iloc` accessor.

```python
weather_data.iloc[:, 3]
```

In [15]:
weather_data.iloc[:, 3]

0         22.9
1         25.1
2         25.7
3         28.0
4         32.3
          ... 
145455    23.4
145456    25.3
145457    26.9
145458    27.0
145459     NaN
Name: MaxTemp, Length: 145460, dtype: float64

**Exercise**

Do you notice something similar to what we learnt in the previous lessons?

<font size="5">[TODO] 📖</font>

It's the slicing operator (`:`).

**Exercise**

Can you select the first 10 rows of the column `MinTemp` from the `weather_data` dataframe?

In [16]:
# [TODO]
weather_data.iloc[:10, 2]

0    13.4
1     7.4
2    12.9
3     9.2
4    17.5
5    14.6
6    14.3
7     7.7
8     9.7
9    13.1
Name: MinTemp, dtype: float64

Let's access the first row of the `weather_data` dataframe using `loc`.

```python
weather_data.loc[0]
```

In [17]:
weather_data.loc[0]

Date             2008-12-01 00:00:00
Location                      Albury
MinTemp                         13.4
MaxTemp                         22.9
Rainfall                         0.6
Evaporation                      NaN
Sunshine                         NaN
WindGustDir                        W
WindGustSpeed                   44.0
WindDir9am                         W
WindDir3pm                       WNW
WindSpeed9am                    20.0
WindSpeed3pm                    24.0
Humidity9am                     71.0
Humidity3pm                     22.0
Pressure9am                   1007.7
Pressure3pm                   1007.1
Cloud9am                         8.0
Cloud3pm                         NaN
Temp9am                         16.9
Temp3pm                         21.8
RainToday                         No
RainTomorrow                      No
Name: 0, dtype: object

It doesn't look that different from `iloc` accessor right? In fact, `iloc` is conceptually simpler than `loc` because it ignores the dataset's indices. When we're using `iloc`, we only need to specify the index of the row(s) and/or column(s) we want to access. `loc` is more flexible because it allows us to specify the label of the row(s) and/or column(s) we want to access.

For example, let's access the first 5 rows from the following columns (`MinTemp`, `MaxTemp`, `RainToday`, `RainTomorrow`) of the `weather_data` dataframe.

```python
weather_data.loc[:4, ["MinTemp", "MaxTemp", "RainToday", "RainTomorrow"]]
```


In [18]:
weather_data.loc[:4, ["MinTemp", "MaxTemp", "RainToday", "RainTomorrow"]]

MinTemp  MaxTemp RainToday RainTomorrow
0     13.4     22.9        No           No
1      7.4     25.1        No           No
2     12.9     25.7        No           No
3      9.2     28.0        No           No
4     17.5     32.3        No           No

Do you notice anything different about `loc` and `iloc` in terms of the provided indices? 

`iloc[:5]` is equivalent to `loc[:4]`. The index in `loc` is inclusive while `iloc` is exclusive.

There is an even **more compact** way to select a subset dataframe from the original dataframe using the `[]` operator. Personally, I prefer to use this method as it is highly readable and easy to use.

Let's do the same using `[]` operator!

```python
weather_data[["MinTemp", "MaxTemp", "RainToday", "RainTomorrow"]].head()
```

In [19]:
weather_data[["MinTemp", "MaxTemp", "RainToday", "RainTomorrow"]].head()

MinTemp  MaxTemp RainToday RainTomorrow
0     13.4     22.9        No           No
1      7.4     25.1        No           No
2     12.9     25.7        No           No
3      9.2     28.0        No           No
4     17.5     32.3        No           No

As shown above, when creating a dataframe, we have an option to specify the `index` of the dataframe instead of using the automatically generated numbers. Thus, in this case, if we want to use `Location` column as the `index` of our `weather_data` dataframe, we could use the `set_index()` method. 

```python
weather_data.set_index("Location")
```

In [20]:
weather_data.set_index("Location")

Date  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
Location                                                                 
Albury   2008-12-01     13.4     22.9       0.6          NaN       NaN   
Albury   2008-12-02      7.4     25.1       0.0          NaN       NaN   
Albury   2008-12-03     12.9     25.7       0.0          NaN       NaN   
Albury   2008-12-04      9.2     28.0       0.0          NaN       NaN   
Albury   2008-12-05     17.5     32.3       1.0          NaN       NaN   
...             ...      ...      ...       ...          ...       ...   
Uluru    2017-06-21      2.8     23.4       0.0          NaN       NaN   
Uluru    2017-06-22      3.6     25.3       0.0          NaN       NaN   
Uluru    2017-06-23      5.4     26.9       0.0          NaN       NaN   
Uluru    2017-06-24      7.8     27.0       0.0          NaN       NaN   
Uluru    2017-06-25     14.9      NaN       0.0          NaN       NaN   

         WindGustDir  WindGustSpeed WindDir9am WindDir3pm  ...  Humidity9am  \
Location                                                   ...                
Albury             W           44.0          W        WNW  ...         71.0   
Albury           WNW           44.0        NNW        WSW  ...         44.0   
Albury           WSW           46.0          W        WSW  ...         38.0   
Albury            NE           24.0         SE          E  ...         45.0   
Albury             W           41.0        ENE         NW  ...         82.0   
...              ...            ...        ...        ...  ...          ...   
Uluru              E           31.0         SE        ENE  ...         51.0   
Uluru            NNW           22.0         SE          N  ...         56.0   
Uluru              N           37.0         SE        WNW  ...         53.0   
Uluru             SE           28.0        SSE          N  ...         51.0   
Uluru            NaN            NaN        ESE        ESE  ...         62.0   

          Humidity3pm  Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  \
Location                                                                       
Albury           22.0       1007.7       1007.1       8.0       NaN     16.9   
Albury           25.0       1010.6       1007.8       NaN       NaN     17.2   
Albury           30.0       1007.6       1008.7       NaN       2.0     21.0   
Albury           16.0       1017.6       1012.8       NaN       NaN     18.1   
Albury           33.0       1010.8       1006.0       7.0       8.0     17.8   
...               ...          ...          ...       ...       ...      ...   
Uluru            24.0       1024.6       1020.3       NaN       NaN     10.1   
Uluru            21.0       1023.5       1019.1       NaN       NaN     10.9   
Uluru            24.0       1021.0       1016.8       NaN       NaN     12.5   
Uluru            24.0       1019.4       1016.5       3.0       2.0     15.1   
Uluru            36.0       1020.2       1017.9       8.0       8.0     15.0   

          Temp3pm  RainToday  RainTomorrow  
Location                                    
Albury       21.8         No            No  
Albury       24.3         No            No  
Albury       23.2         No            No  
Albury       26.5         No            No  
Albury       29.7         No            No  
...           ...        ...           ...  
Uluru        22.4         No            No  
Uluru        24.5         No            No  
Uluru        26.1         No            No  
Uluru        26.0         No            No  
Uluru        20.9         No           NaN  

[145460 rows x 22 columns]

## Selecting

This will be extremely similar to how we filter data in a `numpy` array. Often in a project, we want to select a subset of the data based on certain conditions to perform some analysis in depth.

For example, let's select all records where the `Location` is `Sydney`.

```python
weather_data[weather_data["Location"] == "Sydney"]
```

In [21]:
weather_data[weather_data["Location"] == "Sydney"]

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
30176 2008-02-01   Sydney     19.5     22.4      15.6          6.2       0.0   
30177 2008-02-02   Sydney     19.5     25.6       6.0          3.4       2.7   
30178 2008-02-03   Sydney     21.6     24.5       6.6          2.4       0.1   
30179 2008-02-04   Sydney     20.2     22.8      18.8          2.2       0.0   
30180 2008-02-05   Sydney     19.7     25.7      77.4          NaN       0.0   
...          ...      ...      ...      ...       ...          ...       ...   
33515 2017-06-21   Sydney      8.6     19.6       0.0          2.0       7.8   
33516 2017-06-22   Sydney      9.3     19.2       0.0          2.0       9.2   
33517 2017-06-23   Sydney      9.4     17.7       0.0          2.4       2.7   
33518 2017-06-24   Sydney     10.1     19.3       0.0          1.4       9.3   
33519 2017-06-25   Sydney      7.6     19.3       0.0          3.4       9.4   

      WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  Humidity3pm  \
30176         NaN            NaN          S  ...        92.0         84.0   
30177         NaN            NaN          W  ...        83.0         73.0   
30178         NaN            NaN        ESE  ...        88.0         86.0   
30179         NaN            NaN        NNE  ...        83.0         90.0   
30180         NaN            NaN        NNE  ...        88.0         74.0   
...           ...            ...        ...  ...         ...          ...   
33515         SSE           37.0          W  ...        73.0         52.0   
33516           W           30.0          W  ...        78.0         53.0   
33517           W           24.0        WNW  ...        85.0         56.0   
33518           W           43.0          W  ...        56.0         35.0   
33519           W           35.0          W  ...        73.0         32.0   

       Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  \
30176       1017.6       1017.4       8.0       8.0     20.7     20.9   
30177       1017.9       1016.4       7.0       7.0     22.4     24.8   
30178       1016.7       1015.6       7.0       8.0     23.5     23.0   
30179       1014.2       1011.8       8.0       8.0     21.4     20.9   
30180       1008.3       1004.8       8.0       8.0     22.5     25.5   
...            ...          ...       ...       ...      ...      ...   
33515       1025.9       1025.3       2.0       2.0     10.5     17.9   
33516       1028.5       1024.6       2.0       2.0     11.0     18.7   
33517       1020.8       1015.0       6.0       6.0     10.2     17.3   
33518       1017.3       1015.1       5.0       2.0     12.4     19.0   
33519       1018.6       1015.4       1.0       1.0      9.4     18.8   

       RainToday  RainTomorrow  
30176        Yes           Yes  
30177        Yes           Yes  
30178        Yes           Yes  
30179        Yes           Yes  
30180        Yes           Yes  
...          ...           ...  
33515         No            No  
33516         No            No  
33517         No            No  
33518         No            No  
33519         No            No  

[3344 rows x 23 columns]

As you can see, the result is a dataframe with only the records where the `Location` is `Sydney`. 

**Exercise**

How many rows are there where the `Location` is `Sydney` and the `RainTomorrow` is `Yes`?

**Hint**: We can use `&` (`ampersand`: on a high level, it's similar to `and`).

In [22]:
# [TODO]
weather_data[(weather_data["Location"] == "Sydney") & (weather_data["RainTomorrow"] == "Yes")].shape

(865, 23)

**Exercise**

How many rows are there where the `Evaporation` is greater than `3` or `RainTomorrow` is `Yes`?

**Hint**: We can use `|` (`pipe`: on a high level, it's similar to `or`).

In [23]:
# [TODO]
weather_data[(weather_data["Evaporation"] > 3) | (weather_data["RainTomorrow"] == "Yes")].shape

(78599, 23)

**Exercise**

Can you select all records where the `Location` is either `Sydney`, `Melbourne`, or `Canberra`?

In [24]:
# [TODO]
weather_data[
    (weather_data["Location"] == "Sydney")
    | (weather_data["Location"] == "Melbourne")
    | (weather_data["Location"] == "Canberra")
]

Date   Location  MinTemp  MaxTemp  Rainfall  Evaporation  \
30176 2008-02-01     Sydney     19.5     22.4      15.6          6.2   
30177 2008-02-02     Sydney     19.5     25.6       6.0          3.4   
30178 2008-02-03     Sydney     21.6     24.5       6.6          2.4   
30179 2008-02-04     Sydney     20.2     22.8      18.8          2.2   
30180 2008-02-05     Sydney     19.7     25.7      77.4          NaN   
...          ...        ...      ...      ...       ...          ...   
70388 2017-06-21  Melbourne      7.8     13.5       0.6          0.6   
70389 2017-06-22  Melbourne      6.7     14.1       0.2          0.6   
70390 2017-06-23  Melbourne      7.4     14.9       0.0          1.6   
70391 2017-06-24  Melbourne      8.3     12.3       1.4          1.8   
70392 2017-06-25  Melbourne      8.6     14.3       0.0          2.8   

       Sunshine WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  \
30176       0.0         NaN            NaN          S  ...        92.0   
30177       2.7         NaN            NaN          W  ...        83.0   
30178       0.1         NaN            NaN        ESE  ...        88.0   
30179       0.0         NaN            NaN        NNE  ...        83.0   
30180       0.0         NaN            NaN        NNE  ...        88.0   
...         ...         ...            ...        ...  ...         ...   
70388       0.0         NNW           24.0          W  ...       100.0   
70389       0.0           N           24.0        NNE  ...        93.0   
70390       6.0         NNW           44.0          N  ...        77.0   
70391       5.6          SW           30.0        WNW  ...        75.0   
70392       3.8          NW           35.0          N  ...        67.0   

       Humidity3pm  Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  \
30176         84.0       1017.6       1017.4       8.0       8.0     20.7   
30177         73.0       1017.9       1016.4       7.0       7.0     22.4   
30178         86.0       1016.7       1015.6       7.0       8.0     23.5   
30179         90.0       1014.2       1011.8       8.0       8.0     21.4   
30180         74.0       1008.3       1004.8       8.0       8.0     22.5   
...            ...          ...          ...       ...       ...      ...   
70388         80.0       1028.8       1028.4       7.0       7.0     11.2   
70389         56.0       1028.2       1024.8       7.0       7.0      7.5   
70390         63.0       1018.8       1013.8       1.0       7.0      9.1   
70391         63.0       1018.7       1017.4       2.0       7.0      9.6   
70392         56.0       1017.4       1015.0       7.0       7.0      9.3   

       Temp3pm  RainToday  RainTomorrow  
30176     20.9        Yes           Yes  
30177     24.8        Yes           Yes  
30178     23.0        Yes           Yes  
30179     20.9        Yes           Yes  
30180     25.5        Yes           Yes  
...        ...        ...           ...  
70388     12.7         No            No  
70389     13.5         No            No  
70390     12.5         No           Yes  
70391     11.2        Yes            No  
70392     13.4         No           Yes  

[9973 rows x 23 columns]

The `|` operator works but it's quite cumbersome for that scenario. There's a better way to do the same thing by using the `pandas` `isin` method. `isin` lets you select data whose value **is in** a list of values.

```python
weather_data[weather_data["Location"].isin(["Sydney", "Melbourne", "Canberra"])]
```

In [25]:
weather_data[weather_data["Location"].isin(["Sydney", "Melbourne", "Canberra"])]

Date   Location  MinTemp  MaxTemp  Rainfall  Evaporation  \
30176 2008-02-01     Sydney     19.5     22.4      15.6          6.2   
30177 2008-02-02     Sydney     19.5     25.6       6.0          3.4   
30178 2008-02-03     Sydney     21.6     24.5       6.6          2.4   
30179 2008-02-04     Sydney     20.2     22.8      18.8          2.2   
30180 2008-02-05     Sydney     19.7     25.7      77.4          NaN   
...          ...        ...      ...      ...       ...          ...   
70388 2017-06-21  Melbourne      7.8     13.5       0.6          0.6   
70389 2017-06-22  Melbourne      6.7     14.1       0.2          0.6   
70390 2017-06-23  Melbourne      7.4     14.9       0.0          1.6   
70391 2017-06-24  Melbourne      8.3     12.3       1.4          1.8   
70392 2017-06-25  Melbourne      8.6     14.3       0.0          2.8   

       Sunshine WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  \
30176       0.0         NaN            NaN          S  ...        92.0   
30177       2.7         NaN            NaN          W  ...        83.0   
30178       0.1         NaN            NaN        ESE  ...        88.0   
30179       0.0         NaN            NaN        NNE  ...        83.0   
30180       0.0         NaN            NaN        NNE  ...        88.0   
...         ...         ...            ...        ...  ...         ...   
70388       0.0         NNW           24.0          W  ...       100.0   
70389       0.0           N           24.0        NNE  ...        93.0   
70390       6.0         NNW           44.0          N  ...        77.0   
70391       5.6          SW           30.0        WNW  ...        75.0   
70392       3.8          NW           35.0          N  ...        67.0   

       Humidity3pm  Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  \
30176         84.0       1017.6       1017.4       8.0       8.0     20.7   
30177         73.0       1017.9       1016.4       7.0       7.0     22.4   
30178         86.0       1016.7       1015.6       7.0       8.0     23.5   
30179         90.0       1014.2       1011.8       8.0       8.0     21.4   
30180         74.0       1008.3       1004.8       8.0       8.0     22.5   
...            ...          ...          ...       ...       ...      ...   
70388         80.0       1028.8       1028.4       7.0       7.0     11.2   
70389         56.0       1028.2       1024.8       7.0       7.0      7.5   
70390         63.0       1018.8       1013.8       1.0       7.0      9.1   
70391         63.0       1018.7       1017.4       2.0       7.0      9.6   
70392         56.0       1017.4       1015.0       7.0       7.0      9.3   

       Temp3pm  RainToday  RainTomorrow  
30176     20.9        Yes           Yes  
30177     24.8        Yes           Yes  
30178     23.0        Yes           Yes  
30179     20.9        Yes           Yes  
30180     25.5        Yes           Yes  
...        ...        ...           ...  
70388     12.7         No            No  
70389     13.5         No            No  
70390     12.5         No           Yes  
70391     11.2        Yes            No  
70392     13.4         No           Yes  

[9973 rows x 23 columns]

Remember that we have seen lots of `NaN` values above. How do we filter out the records with missing values? We can use the `pandas` `isnull` method. The opposite of `isnull` is `notnull`. We can also use `~` (`tilde`) to invert the result.

```python
weather_data[weather_data["Evaporation"].isnull()]
weather_data[~weather_data["Evaporation"].isnull()]
```

In [26]:
weather_data[weather_data["Evaporation"].isnull()]

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0      2008-12-01   Albury     13.4     22.9       0.6          NaN       NaN   
1      2008-12-02   Albury      7.4     25.1       0.0          NaN       NaN   
2      2008-12-03   Albury     12.9     25.7       0.0          NaN       NaN   
3      2008-12-04   Albury      9.2     28.0       0.0          NaN       NaN   
4      2008-12-05   Albury     17.5     32.3       1.0          NaN       NaN   
...           ...      ...      ...      ...       ...          ...       ...   
145455 2017-06-21    Uluru      2.8     23.4       0.0          NaN       NaN   
145456 2017-06-22    Uluru      3.6     25.3       0.0          NaN       NaN   
145457 2017-06-23    Uluru      5.4     26.9       0.0          NaN       NaN   
145458 2017-06-24    Uluru      7.8     27.0       0.0          NaN       NaN   
145459 2017-06-25    Uluru     14.9      NaN       0.0          NaN       NaN   

       WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  Humidity3pm  \
0                W           44.0          W  ...        71.0         22.0   
1              WNW           44.0        NNW  ...        44.0         25.0   
2              WSW           46.0          W  ...        38.0         30.0   
3               NE           24.0         SE  ...        45.0         16.0   
4                W           41.0        ENE  ...        82.0         33.0   
...            ...            ...        ...  ...         ...          ...   
145455           E           31.0         SE  ...        51.0         24.0   
145456         NNW           22.0         SE  ...        56.0         21.0   
145457           N           37.0         SE  ...        53.0         24.0   
145458          SE           28.0        SSE  ...        51.0         24.0   
145459         NaN            NaN        ESE  ...        62.0         36.0   

        Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  \
0            1007.7       1007.1       8.0       NaN     16.9     21.8   
1            1010.6       1007.8       NaN       NaN     17.2     24.3   
2            1007.6       1008.7       NaN       2.0     21.0     23.2   
3            1017.6       1012.8       NaN       NaN     18.1     26.5   
4            1010.8       1006.0       7.0       8.0     17.8     29.7   
...             ...          ...       ...       ...      ...      ...   
145455       1024.6       1020.3       NaN       NaN     10.1     22.4   
145456       1023.5       1019.1       NaN       NaN     10.9     24.5   
145457       1021.0       1016.8       NaN       NaN     12.5     26.1   
145458       1019.4       1016.5       3.0       2.0     15.1     26.0   
145459       1020.2       1017.9       8.0       8.0     15.0     20.9   

        RainToday  RainTomorrow  
0              No            No  
1              No            No  
2              No            No  
3              No            No  
4              No            No  
...           ...           ...  
145455         No            No  
145456         No            No  
145457         No            No  
145458         No            No  
145459         No           NaN  

[62790 rows x 23 columns]

In [27]:
weather_data[~weather_data["Evaporation"].isnull()]

Date   Location  MinTemp  MaxTemp  Rainfall  Evaporation  \
6049   2009-01-01      Cobar     17.9     35.2       0.0         12.0   
6050   2009-01-02      Cobar     18.4     28.9       0.0         14.8   
6051   2009-01-03      Cobar     15.5     34.1       0.0         12.6   
6052   2009-01-04      Cobar     19.4     37.6       0.0         10.8   
6053   2009-01-05      Cobar     21.9     38.4       0.0         11.4   
...           ...        ...      ...      ...       ...          ...   
143876 2017-06-20  Katherine     15.0     32.0       0.0          9.6   
143877 2017-06-21  Katherine     15.4     31.5       0.0          3.6   
143878 2017-06-22  Katherine     12.2     31.5       0.0          4.0   
143879 2017-06-23  Katherine     14.9     31.5       0.0          6.4   
143880 2017-06-24  Katherine     14.0      NaN       0.0         10.0   

        Sunshine WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  \
6049        12.3         SSW           48.0        ENE  ...        20.0   
6050        13.0           S           37.0        SSE  ...        30.0   
6051        13.3          SE           30.0        NaN  ...         NaN   
6052        10.6         NNE           46.0        NNE  ...        42.0   
6053        12.2         WNW           31.0        WNW  ...        37.0   
...          ...         ...            ...        ...  ...         ...   
143876       NaN           E           35.0         SE  ...        47.0   
143877       NaN         ESE           43.0        ESE  ...        70.0   
143878       NaN         ESE           37.0         SE  ...        45.0   
143879       NaN         ESE           41.0         SE  ...        74.0   
143880       NaN           E           35.0          E  ...        74.0   

        Humidity3pm  Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  \
6049           13.0       1006.3       1004.4       2.0       5.0     26.6   
6050            8.0       1012.9       1012.1       1.0       1.0     20.3   
6051            7.0          NaN       1011.6       NaN       1.0      NaN   
6052           22.0       1012.3       1009.2       1.0       6.0     28.7   
6053           22.0       1012.7       1009.1       1.0       5.0     29.1   
...             ...          ...          ...       ...       ...      ...   
143876          NaN       1015.9       1012.1       0.0       NaN     21.8   
143877          NaN       1016.6       1012.7       0.0       NaN     21.6   
143878          NaN       1017.1       1012.8       0.0       NaN     19.9   
143879          NaN       1016.1       1011.8       0.0       NaN     22.8   
143880          NaN       1014.6       1011.1       2.0       NaN     22.2   

        Temp3pm  RainToday  RainTomorrow  
6049       33.4         No            No  
6050       27.0         No            No  
6051       32.7         No            No  
6052       34.9         No            No  
6053       35.6         No            No  
...         ...        ...           ...  
143876      NaN         No            No  
143877      NaN         No            No  
143878      NaN         No            No  
143879      NaN         No            No  
143880      NaN         No           NaN  

[82670 rows x 23 columns]

**Exercise**

Can you check if the sum of rows of the 2 dataframes we got above is the same as the number of rows in the `weather_data` dataframe?

In [28]:
# [TODO]
weather_data[weather_data["Evaporation"].isnull()].shape[0] + weather_data[~weather_data["Evaporation"].isnull()].shape[0] == weather_data.shape[0]

True

## Assigning

Assigning data to a dataframe is very simple and very similar to how we add a new `key-value` pair to an existing dictionary.

**Exercise**

Can you create a column called `MinTemp > 5` to indicate whether the `MinTemp` is greater than 5 or not? The column will contain value "Yes" if the `MinTemp` is greater than 5, and "No" otherwise.

**Hint**: Use `np.where()`.

In [29]:
# [TODO]
import numpy as np

weather_data["MinTemp > 5"] = np.where(weather_data["MinTemp"] > 5, "Yes", "No")

# Summary Functions and Maps

## Summary Functions
`pandas` has functions like `describe()` and `info()` which provide a summary of the data. They can be used to give you an overview of the data by displaying relevant descriptive statistics of the data.

Let's see how `describe()` works!

Note that when the dataframe has too many columns that can't be easily displayed in a single screen, we can transpose the result to view it better.

```python
weather_data.describe().T
```

In [30]:
weather_data.describe().T

count         mean        std    min     25%     50%  \
MinTemp        143975.0    12.194034   6.398495   -8.5     7.6    12.0   
MaxTemp        144199.0    23.221348   7.119049   -4.8    17.9    22.6   
Rainfall       142199.0     2.360918   8.478060    0.0     0.0     0.0   
Evaporation     82670.0     5.468232   4.193704    0.0     2.6     4.8   
Sunshine        75625.0     7.611178   3.785483    0.0     4.8     8.4   
WindGustSpeed  135197.0    40.035230  13.607062    6.0    31.0    39.0   
WindSpeed9am   143693.0    14.043426   8.915375    0.0     7.0    13.0   
WindSpeed3pm   142398.0    18.662657   8.809800    0.0    13.0    19.0   
Humidity9am    142806.0    68.880831  19.029164    0.0    57.0    70.0   
Humidity3pm    140953.0    51.539116  20.795902    0.0    37.0    52.0   
Pressure9am    130395.0  1017.649940   7.106530  980.5  1012.9  1017.6   
Pressure3pm    130432.0  1015.255889   7.037414  977.1  1010.4  1015.2   
Cloud9am        89572.0     4.447461   2.887159    0.0     1.0     5.0   
Cloud3pm        86102.0     4.509930   2.720357    0.0     2.0     5.0   
Temp9am        143693.0    16.990631   6.488753   -7.2    12.3    16.7   
Temp3pm        141851.0    21.683390   6.936650   -5.4    16.6    21.1   

                  75%     max  
MinTemp          16.9    33.9  
MaxTemp          28.2    48.1  
Rainfall          0.8   371.0  
Evaporation       7.4   145.0  
Sunshine         10.6    14.5  
WindGustSpeed    48.0   135.0  
WindSpeed9am     19.0   130.0  
WindSpeed3pm     24.0    87.0  
Humidity9am      83.0   100.0  
Humidity3pm      66.0   100.0  
Pressure9am    1022.4  1041.0  
Pressure3pm    1020.0  1039.6  
Cloud9am          7.0     9.0  
Cloud3pm          7.0     9.0  
Temp9am          21.6    40.2  
Temp3pm          26.4    46.7

`describe()` gives you the following information.
- `count`: the number of `notnull` rows for that particular column.
- `mean`: the mean of the column.
- `std`: the standard deviation of the column.
- `min`: the minimum value of the column.
- `25%`: the 25th percentile of the column.
- `50%`: the 50th percentile of the column.
- `75%`: the 75th percentile of the column.
- `max`: the maximum value of the column.

If you use `describe()` as is, you will only get summary statistics for the **numerical** columns. If you want to get the summary statistics for all columns, you can use the `.describe(include="all")` method.

```python
weather_data.describe(include="all").T
```

In [31]:
weather_data.describe(include="all").T

/tmp/ipykernel_24269/2476468055.py:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  weather_data.describe(include="all").T


count unique                  top    freq      first  \
Date             145460   3436  2013-11-12 00:00:00      49 2007-11-01   
Location         145460     49             Canberra    3436        NaT   
MinTemp        143975.0    NaN                  NaN     NaN        NaT   
MaxTemp        144199.0    NaN                  NaN     NaN        NaT   
Rainfall       142199.0    NaN                  NaN     NaN        NaT   
Evaporation     82670.0    NaN                  NaN     NaN        NaT   
Sunshine        75625.0    NaN                  NaN     NaN        NaT   
WindGustDir      135134     16                    W    9915        NaT   
WindGustSpeed  135197.0    NaN                  NaN     NaN        NaT   
WindDir9am       134894     16                    N   11758        NaT   
WindDir3pm       141232     16                   SE   10838        NaT   
WindSpeed9am   143693.0    NaN                  NaN     NaN        NaT   
WindSpeed3pm   142398.0    NaN                  NaN     NaN        NaT   
Humidity9am    142806.0    NaN                  NaN     NaN        NaT   
Humidity3pm    140953.0    NaN                  NaN     NaN        NaT   
Pressure9am    130395.0    NaN                  NaN     NaN        NaT   
Pressure3pm    130432.0    NaN                  NaN     NaN        NaT   
Cloud9am        89572.0    NaN                  NaN     NaN        NaT   
Cloud3pm        86102.0    NaN                  NaN     NaN        NaT   
Temp9am        143693.0    NaN                  NaN     NaN        NaT   
Temp3pm        141851.0    NaN                  NaN     NaN        NaT   
RainToday        142199      2                   No  110319        NaT   
RainTomorrow     142193      2                   No  110316        NaT   
MinTemp > 5      145460      2                  Yes  124389        NaT   

                    last         mean        std    min     25%     50%  \
Date          2017-06-25          NaN        NaN    NaN     NaN     NaN   
Location             NaT          NaN        NaN    NaN     NaN     NaN   
MinTemp              NaT    12.194034   6.398495   -8.5     7.6    12.0   
MaxTemp              NaT    23.221348   7.119049   -4.8    17.9    22.6   
Rainfall             NaT     2.360918    8.47806    0.0     0.0     0.0   
Evaporation          NaT     5.468232   4.193704    0.0     2.6     4.8   
Sunshine             NaT     7.611178   3.785483    0.0     4.8     8.4   
WindGustDir          NaT          NaN        NaN    NaN     NaN     NaN   
WindGustSpeed        NaT     40.03523  13.607062    6.0    31.0    39.0   
WindDir9am           NaT          NaN        NaN    NaN     NaN     NaN   
WindDir3pm           NaT          NaN        NaN    NaN     NaN     NaN   
WindSpeed9am         NaT    14.043426   8.915375    0.0     7.0    13.0   
WindSpeed3pm         NaT    18.662657     8.8098    0.0    13.0    19.0   
Humidity9am          NaT    68.880831  19.029164    0.0    57.0    70.0   
Humidity3pm          NaT    51.539116  20.795902    0.0    37.0    52.0   
Pressure9am          NaT   1017.64994    7.10653  980.5  1012.9  1017.6   
Pressure3pm          NaT  1015.255889   7.037414  977.1  1010.4  1015.2   
Cloud9am             NaT     4.447461   2.887159    0.0     1.0     5.0   
Cloud3pm             NaT      4.50993   2.720357    0.0     2.0     5.0   
Temp9am              NaT    16.990631   6.488753   -7.2    12.3    16.7   
Temp3pm              NaT     21.68339    6.93665   -5.4    16.6    21.1   
RainToday            NaT          NaN        NaN    NaN     NaN     NaN   
RainTomorrow         NaT          NaN        NaN    NaN     NaN     NaN   
MinTemp > 5          NaT          NaN        NaN    NaN     NaN     NaN   

                  75%     max  
Date              NaN     NaN  
Location          NaN     NaN  
MinTemp          16.9    33.9  
MaxTemp          28.2    48.1  
Rainfall          0.8   371.0  
Evaporation       7.4   145.0  
Sunshine         10.6    14.5  
WindGustDir       NaN     NaN  
Win

It's actually extremely hard to read, so normally, we view the summary statistics of **numerical** and **categorical** columns separately.

To view the summary statistics of only **categorical** columns, we can use the `.describe(include="object")` method.
- `count`: the number of `notnull` rows for that particular column.
- `unique`: the number of unique values in the column.
- `top`: the top most common values in the column.
- `freq`: the frequency of the top most common values.

```python
weather_data.describe(include="object").T
```

In [32]:
weather_data.describe(include="object").T

count unique       top    freq
Location      145460     49  Canberra    3436
WindGustDir   135134     16         W    9915
WindDir9am    134894     16         N   11758
WindDir3pm    141232     16        SE   10838
RainToday     142199      2        No  110319
RainTomorrow  142193      2        No  110316
MinTemp > 5   145460      2       Yes  124389

Let's see how `info()` works!
- `Non-Null Count`: the number of `notnull` rows for that particular column.
- `Dtype`: the data types of the columns.

`info()` gives us an idea to see whether the `pd.read_csv()` method used before has inferred the data types of the columns correctly. We can easily assign the correct data types to each column while reading the data.

```python
weather_data.info()
```

In [33]:
weather_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 24 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   Date           145460 non-null  datetime64[ns]
 1   Location       145460 non-null  object        
 2   MinTemp        143975 non-null  float64       
 3   MaxTemp        144199 non-null  float64       
 4   Rainfall       142199 non-null  float64       
 5   Evaporation    82670 non-null   float64       
 6   Sunshine       75625 non-null   float64       
 7   WindGustDir    135134 non-null  object        
 8   WindGustSpeed  135197 non-null  float64       
 9   WindDir9am     134894 non-null  object        
 10  WindDir3pm     141232 non-null  object        
 11  WindSpeed9am   143693 non-null  float64       
 12  WindSpeed3pm   142398 non-null  float64       
 13  Humidity9am    142806 non-null  float64       
 14  Humidity3pm    140953 non-null  float64       
 15  

To view the data type of a single column, we can use the `.dtype` attribute.

```python
weather_data["Location"].dtype
```

In [34]:
weather_data["Location"].dtype

dtype('O')

To quickly view the data types of each column, we can use the `.dtypes` attribute.

```python
weather_data.dtypes
```

In [35]:
weather_data.dtypes

Date             datetime64[ns]
Location                 object
MinTemp                 float64
MaxTemp                 float64
Rainfall                float64
Evaporation             float64
Sunshine                float64
WindGustDir              object
WindGustSpeed           float64
WindDir9am               object
WindDir3pm               object
WindSpeed9am            float64
WindSpeed3pm            float64
Humidity9am             float64
Humidity3pm             float64
Pressure9am             float64
Pressure3pm             float64
Cloud9am                float64
Cloud3pm                float64
Temp9am                 float64
Temp3pm                 float64
RainToday                object
RainTomorrow             object
MinTemp > 5              object
dtype: object

Regarding unique values in a column, we can use:
- `unique()` to get the list of unique values in the column of interest.
- `nunique()` to get the number of unique values in the column of interest.
- `value_counts()` to get the frequency of each unique value in the column of interest.

```python
weather_data["Location"].unique()
weather_data["Location"].nunique()
weather_data["Location"].value_counts()
```

In [36]:
weather_data["Location"].unique()

array(['Albury', 'BadgerysCreek', 'Cobar', 'CoffsHarbour', 'Moree',
       'Newcastle', 'NorahHead', 'NorfolkIsland', 'Penrith', 'Richmond',
       'Sydney', 'SydneyAirport', 'WaggaWagga', 'Williamtown',
       'Wollongong', 'Canberra', 'Tuggeranong', 'MountGinini', 'Ballarat',
       'Bendigo', 'Sale', 'MelbourneAirport', 'Melbourne', 'Mildura',
       'Nhil', 'Portland', 'Watsonia', 'Dartmoor', 'Brisbane', 'Cairns',
       'GoldCoast', 'Townsville', 'Adelaide', 'MountGambier', 'Nuriootpa',
       'Woomera', 'Albany', 'Witchcliffe', 'PearceRAAF', 'PerthAirport',
       'Perth', 'SalmonGums', 'Walpole', 'Hobart', 'Launceston',
       'AliceSprings', 'Darwin', 'Katherine', 'Uluru'], dtype=object)

In [37]:
weather_data["Location"].nunique()

49

In [38]:
weather_data["Location"].value_counts()

Canberra            3436
Sydney              3344
Darwin              3193
Melbourne           3193
Brisbane            3193
Adelaide            3193
Perth               3193
Hobart              3193
Albany              3040
MountGambier        3040
Ballarat            3040
Townsville          3040
GoldCoast           3040
Cairns              3040
Launceston          3040
AliceSprings        3040
Bendigo             3040
Albury              3040
MountGinini         3040
Wollongong          3040
Newcastle           3039
Tuggeranong         3039
Penrith             3039
Woomera             3009
Nuriootpa           3009
Cobar               3009
CoffsHarbour        3009
Moree               3009
Sale                3009
PerthAirport        3009
PearceRAAF          3009
Witchcliffe         3009
BadgerysCreek       3009
Mildura             3009
NorfolkIsland       3009
MelbourneAirport    3009
Richmond            3009
SydneyAirport       3009
WaggaWagga          3009
Williamtown         3009


## Maps

**Mapping** is a way to map one set of values to another set of values. This is one of the way to transform the raw data to the desired format. This is something that data scientists do a lot.

There are lots of mapping methods in `pandas`, but you'll often use the `map()` method.

In data science, it's very common to standardise features to a common scale. One of the most common scaler is called **Standard Scaler** where we transform each data point of the feature into the `z-score`. 

z = $\displaystyle \frac{x - mean}{std}$

Let's see how we normalise the `Rainfall` column using `map()` method! We'll also save the normalised data to a new column called `NormalisedRainfall`.

```python
mean_rainfall = weather_data["Rainfall"].mean()
std_rainfall = weather_data["Rainfall"].std()

weather_data["NormalisedRainfall"] = weather_data["Rainfall"].map(
    lambda x: (x - mean_rainfall) / std_rainfall
)
```

In [39]:
mean_rainfall = weather_data["Rainfall"].mean()
std_rainfall = weather_data["Rainfall"].std()

weather_data["NormalisedRainfall"] = weather_data["Rainfall"].map(
    lambda x: (x - mean_rainfall) / std_rainfall
)

Notice the usage of the anonymous (`lambda`) function. The `lambda` function passed to `map()` expects a single value from the Series (a point value, in the above example), and return a transformed version of that value. `map()` returns a new Series where all the values have been transformed by the anonymous function.

Let's look at the normalised column!

In [40]:
weather_data["NormalisedRainfall"].head()

0   -0.207703
1   -0.278474
2   -0.278474
3   -0.278474
4   -0.160522
Name: NormalisedRainfall, dtype: float64

In [41]:
weather_data["NormalisedRainfall"].describe()

count    1.421990e+05
mean    -3.473738e-14
std      1.000000e+00
min     -2.784739e-01
25%     -2.784739e-01
50%     -2.784739e-01
75%     -1.841127e-01
max      4.348154e+01
Name: NormalisedRainfall, dtype: float64

In fact, we don't even need to use `map()` to normalise the data. There are plenty of common mapping operations already built-in `pandas`.

Let's do the same thing in a faster way!

```python
mean_rainfall = weather_data["Rainfall"].mean()
std_rainfall = weather_data["Rainfall"].std()

weather_data["NormalisedRainfall_v2"] = (weather_data["Rainfall"] - mean_rainfall)/std_rainfall
```

In [42]:
weather_data["NormalisedRainfall_v2"] = (weather_data["Rainfall"] - mean_rainfall)/std_rainfall

In [43]:
weather_data["NormalisedRainfall_v2"].describe()

count    1.421990e+05
mean    -3.473738e-14
std      1.000000e+00
min     -2.784739e-01
25%     -2.784739e-01
50%     -2.784739e-01
75%     -1.841127e-01
max      4.348154e+01
Name: NormalisedRainfall_v2, dtype: float64

We can see that `NormalisedRainfall_v2` is exactly the same as `NormalisedRainfall`. In fact, `NormalisedRainfall_v2` is even calculated faster. Let's verify it by using `%timeit` magic command.

```python
%timeit weather_data["NormalisedRainfall"] = weather_data["Rainfall"].map(lambda x: (x - mean_rainfall) / std_rainfall)
%timeit weather_data["NormalisedRainfall_v2"] = (weather_data["Rainfall"] - mean_rainfall)/std_rainfall
```

In [44]:
%timeit weather_data["NormalisedRainfall"] = weather_data["Rainfall"].map(lambda x: (x - mean_rainfall) / std_rainfall)

34.2 ms ± 541 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [45]:
%timeit weather_data["NormalisedRainfall_v2"] = (weather_data["Rainfall"] - mean_rainfall)/std_rainfall

824 µs ± 37.7 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


Despite this, `map()` method is still very useful as it is a lot more flexible than the built-in operations. You can use `map()` to apply conditional logics, which can't be done using the built-in operations.

There is another very common mapping method called `apply()`. It is similar to `map()` but it works on a DataFrame instead of a Series.

Let's say we want to apply the same standardisation to the following columns ("Rainfall", "Temp9am", "WindGustSpeed") in the `weather_data` dataframe!

```python
df = weather_data[["Rainfall", "Temp9am", "WindGustSpeed"]]

def standardise(df):
    mean_column = df.mean()
    std_column = df.std()
    return (df - mean_column) / std_column

standardised_df = df.apply(standardise, axis=0)
```

In [46]:
df = weather_data[["Rainfall", "Temp9am", "WindGustSpeed"]]

def standardise(df):
    mean_column = df.mean()
    std_column = df.std()
    return (df - mean_column) / std_column

standardised_df = df.apply(standardise, axis=0)

Let's compare `weather_data["NormalisedRainfall"]` and `standardised_df["Rainfall"]` to see that we have the same result!

To compare 2 `pd.Series`, we can use the `equals()` method. More about the `equals()` method is available [here](https://pandas.pydata.org/docs/reference/api/pandas.Series.equals.html).

In [47]:
weather_data["NormalisedRainfall"].equals(standardised_df["Rainfall"])

True

In [48]:
weather_data["NormalisedRainfall"].iloc[100:105]

100   -0.136932
101   -0.278474
102    0.405645
103    0.075381
104    1.089764
Name: NormalisedRainfall, dtype: float64

In [49]:
standardised_df["Rainfall"].iloc[100:105]

100   -0.136932
101   -0.278474
102    0.405645
103    0.075381
104    1.089764
Name: Rainfall, dtype: float64

# Grouping and Sorting

## Grouping
It is very common in data science that we need to aggregate the data based on some criteria and perform some calculations on grouped data. In order to do so, we can use the `groupby()` method.

How `groupby()` works is as follows:

![image](../../images/pd_groupby.png)

Let's find the total `Rainfall` for each `Location` in the `weather_data` dataframe! We'll assign the result to a variable called `total_rainfall_by_location`.

```python
total_rainfall_by_location = weather_data.groupby(["Location"])["Rainfall"].sum()
```

In [50]:
total_rainfall_by_location = weather_data.groupby(["Location"])["Rainfall"].sum()

In [51]:
total_rainfall_by_location

Location
Adelaide             4841.6
Albany               6827.8
Albury               5763.4
AliceSprings         2676.8
BadgerysCreek        6421.4
Ballarat             5268.8
Bendigo              4913.2
Brisbane             9941.0
Cairns              17157.2
Canberra             5953.2
Cobar                3368.4
CoffsHarbour        14946.6
Dartmoor             6315.2
Darwin              16260.2
GoldCoast           11232.8
Hobart               5106.6
Katherine            4993.7
Launceston           6092.3
Melbourne            4553.6
MelbourneAirport     4369.0
Mildura              2841.8
Moree                4652.6
MountGambier         6327.4
MountGinini          9570.6
Newcastle            9408.4
Nhil                 1466.8
NorahHead            9921.4
NorfolkIsland        9270.4
Nuriootpa            4175.2
PearceRAAF           4610.0
Penrith              6447.6
Perth                6086.8
PerthAirport         5300.8
Portland             7581.0
Richmond             6310.6
Sale       

In [52]:
type(total_rainfall_by_location)

pandas.core.series.Series

So the result is a `Series` with the `Location` as the index and the total `Rainfall` as the value. Often, we want to have the result as a dataframe instead of a `Series`. In order to do that, we can use the `reset_index()` method, which will reset the index of the result `Series` and allow us to have `Location` as another column in the result dataframe.

```python
total_rainfall_by_location = total_rainfall_by_location.reset_index()
```

In [53]:
total_rainfall_by_location = total_rainfall_by_location.reset_index()
total_rainfall_by_location.head()

Location  Rainfall
0       Adelaide    4841.6
1         Albany    6827.8
2         Albury    5763.4
3   AliceSprings    2676.8
4  BadgerysCreek    6421.4

We could combine everything in 1 single line as follows:

```python
total_rainfall_by_location = weather_data.groupby(["Location"])["Rainfall"].sum().reset_index()
```

In [54]:
total_rainfall_by_location = weather_data.groupby(["Location"])["Rainfall"].sum().reset_index()
total_rainfall_by_location.head()

Location  Rainfall
0       Adelaide    4841.6
1         Albany    6827.8
2         Albury    5763.4
3   AliceSprings    2676.8
4  BadgerysCreek    6421.4

**Exercise**

1. Can you find the maximum `MinTemp` for each location in the `weather_data` dataframe?
1. Based on the result above, can you find the location with the highest `MinTemp`?

**Hint**: You can use the `max()` method to find the maximum value in a `Series`.

In [55]:
# [TODO]
# find maximum MinTemp for each location
result = weather_data.groupby(["Location"])["MinTemp"].max().reset_index()

# find the maximum MinTemp from the result above
max_min_temp = result["MinTemp"].max()

# find the location with the maximum MinTemp
result[result["MinTemp"] == max_min_temp]

Location  MinTemp
0  Adelaide     33.9

There are many aggregate functions in `pandas`. The most common ones that you'll use are: `sum()`, `mean()`, `min()`, `max()`, `median()`. If you want to perform multiple aggregate functions on a single column, you can use the `agg()` method.

For example, we can write 1 single line of code to find the following information about the `Rainfall` column:
- Total `Rainfall` per `Location`.
- Average `Rainfall` per `Location`.
- Minimum `Rainfall` per `Location`.
- Maximum `Rainfall` per `Location`.

```python
weather_data.groupby(["Location"])["Rainfall"].agg(["sum", "mean", "min", "max"]).reset_index()
```

In [56]:
result = weather_data.groupby(["Location"])["Rainfall"].agg(["sum", "mean", "min", "max"]).reset_index()
result.head()

Location     sum      mean  min    max
0       Adelaide  4841.6  1.566354  0.0   75.2
1         Albany  6827.8  2.263859  0.0   49.4
2         Albury  5763.4  1.914115  0.0  104.2
3   AliceSprings  2676.8  0.882850  0.0   62.0
4  BadgerysCreek  6421.4  2.193101  0.0  116.0

Notice that the column names are automatically generated by the `agg()` method. Usually, it's a good idea to rename the columns to make them more readable. This is because if we perform the same set of functions on another column (e.g. `MaxTemp`), the column names will be the same as the aggregated functions, and it will be hard to combine the results and make sense of the data.

We can use the `rename()` method to rename dataframe columns. 
- It's a very useful method as we can rename multiple columns at once by supplying a dictionary of old and new column names to the method. 
- We can choose to have the operation done `inplace` or `not inplace`.
    - `inplace` means we will be saving the new result in the original dataframe.
    - `not inplace` means we will be creating a new dataframe with the new result.

Let's change the column names `not inplace` first and understand what it does.

```python
result.rename(columns={
    "sum": "TotalRainfall",
    "mean": "AverageRainfall",
    "min": "MinRainfall",
    "max": "MaxRainfall"
}, inplace=False)
```

In [57]:
result.rename(columns={
    "sum": "TotalRainfall",
    "mean": "AverageRainfall",
    "min": "MinRainfall",
    "max": "MaxRainfall"
}, inplace=False).head()

Location  TotalRainfall  AverageRainfall  MinRainfall  MaxRainfall
0       Adelaide         4841.6         1.566354          0.0         75.2
1         Albany         6827.8         2.263859          0.0         49.4
2         Albury         5763.4         1.914115          0.0        104.2
3   AliceSprings         2676.8         0.882850          0.0         62.0
4  BadgerysCreek         6421.4         2.193101          0.0        116.0

If we access `result` variable, we can see that it remains intact.

In [58]:
result.head()

Location     sum      mean  min    max
0       Adelaide  4841.6  1.566354  0.0   75.2
1         Albany  6827.8  2.263859  0.0   49.4
2         Albury  5763.4  1.914115  0.0  104.2
3   AliceSprings  2676.8  0.882850  0.0   62.0
4  BadgerysCreek  6421.4  2.193101  0.0  116.0

Let's change the column names `inplace` this time!

```python
result.rename(columns={
    "sum": "TotalRainfall",
    "mean": "AverageRainfall",
    "min": "MinRainfall",
    "max": "MaxRainfall"
}, inplace=True)
```

In [59]:
result.rename(columns={
    "sum": "TotalRainfall",
    "mean": "AverageRainfall",
    "min": "MinRainfall",
    "max": "MaxRainfall"
}, inplace=True)

We can see that `result` dataframe is now changed with the new column names.

In [60]:
result.head()

Location  TotalRainfall  AverageRainfall  MinRainfall  MaxRainfall
0       Adelaide         4841.6         1.566354          0.0         75.2
1         Albany         6827.8         2.263859          0.0         49.4
2         Albury         5763.4         1.914115          0.0        104.2
3   AliceSprings         2676.8         0.882850          0.0         62.0
4  BadgerysCreek         6421.4         2.193101          0.0        116.0

In the example thus far, we have only grouped by 1 single column (e.g. `Location`). In fact, we can group by multiple columns as well.

Remember that we have created a column called `MinTemp > 5` above which has the value of `Yes` if the `MinTemp` is greater than 5 and `No` otherwise.

Let's find the total `Rainfall` for each `Location` and `MinTemp > 5` in the `weather_data` dataframe! We'll assign the result to a variable called `result`.

```python
result = weather_data.groupby(["Location", "MinTemp > 5"])["Rainfall"].sum().reset_index()
```

In [61]:
result = weather_data.groupby(["Location", "MinTemp > 5"])["Rainfall"].sum().reset_index()
result.head()

Location MinTemp > 5  Rainfall
0  Adelaide          No     103.8
1  Adelaide         Yes    4737.8
2    Albany          No      96.8
3    Albany         Yes    6731.0
4    Albury          No     552.6

This simple aggregation can already tell us **some interesting information about the data**. You can see that in any `Location` if `MinTemp` is greater than 5, there's a higher chance that it will rain (`Rainfall` is higher) than if `MinTemp` is less than or equal to 5.

I'm going to perform some operations on the `Date` column of the `weather_data` dataframe to extract the `Month` and `Year` from the `Date` column. This is the reason why we need to be very careful when reading in `datetime` information. Most of the time, we want to ensure that the `Date` column is recognised as actual `datetime` instead of a string.

```python
weather_data["Month"] = weather_data["Date"].dt.month
weather_data["Year"] = weather_data["Date"].dt.year
```

In [62]:
weather_data["Month"] = weather_data["Date"].dt.month
weather_data["Year"] = weather_data["Date"].dt.year

Working with `datetime` is fun but a bit tricky. Don't worry about it too much at the moment! 

[Here](https://www.dataquest.io/blog/datetime-in-pandas/) is a pretty good tutorial on how to start working with `datetime` in `pandas`.

**Exercise**

Can you view 10 random samples `["Date", "Month", "Year"]` from the `weather_data` dataframe to see if we have performed the operations correctly?

In [63]:
# [TODO]
weather_data[["Date", "Month", "Year"]].sample(10)

Date  Month  Year
38229  2013-11-25     11  2013
122981 2015-02-27      2  2015
42650  2009-03-14      3  2009
77     2009-02-16      2  2009
77272  2015-07-10      7  2015
74011  2014-10-31     10  2014
136644 2010-06-27      6  2010
118710 2012-01-17      1  2012
28779  2013-08-29      8  2013
34055  2010-06-20      6  2010

**Exercise**

Can you tell me which `Month` has the highest `Rainfall` in `Sydney`?

**Hint**: You need to filter for `Location == Sydney` first.

In [64]:
# [TODO]
weather_data[weather_data["Location"] == "Sydney"].groupby(["Month"])["Rainfall"].sum().reset_index()

Month  Rainfall
0       1     872.6
1       2    1102.6
2       3    1390.0
3       4    1402.6
4       5     754.6
5       6    1715.2
6       7     796.0
7       8     601.6
8       9     489.0
9      10     601.6
10     11     775.0
11     12     593.2

**Exercise**

How many `Date`s are there for each `Year` for `Sydney`?

In [65]:
# [TODO]
weather_data[weather_data["Location"] == "Sydney"].groupby(["Year"])["Date"].count().reset_index()

Year  Date
0  2008   335
1  2009   365
2  2010   365
3  2011   335
4  2012   335
5  2013   337
6  2014   365
7  2015   365
8  2016   366
9  2017   176

We can see that there're some gaps in the data. For instance, `2008`, `2011`, `2012`, and `2013` didn't contain 1-full year worth of data. `2017` is an exception because we already know the data is only collected until `Jun 2017`.

**Exercise**

Can you tell me which `Year` has the highest amount of `Rainfall` in `Adelaide`?

In [66]:
# [TODO]
weather_data[weather_data["Location"] == "Adelaide"].groupby(["Year"])["Rainfall"].sum().reset_index()

Year  Rainfall
0  2008     235.2
1  2009     517.2
2  2010     592.6
3  2011     528.0
4  2012     516.4
5  2013     495.6
6  2014     534.2
7  2015     395.2
8  2016     820.8
9  2017     206.4

## Sorting

Sorting is a very common operation in data/programming. We can sort a `Series` or a `DataFrame` by using the `sort_values()` method.

How sorting works together with `groupby()` is as follows:

![image](../../images/pd_groupby_sort.png)

Let's revisit the `total_rainfall_by_location` dataframe we created earlier. Let's try to sort that dataframe by `Rainfall` in ascending order.

```python
total_rainfall_by_location = weather_data.groupby(["Location"])["Rainfall"].sum().reset_index()
total_rainfall_by_location = total_rainfall_by_location.sort_values(by="Rainfall", ascending=True)
```

In [67]:
total_rainfall_by_location = weather_data.groupby(["Location"])["Rainfall"].sum().reset_index()

print("Before sorting: ")
total_rainfall_by_location

Before sorting: 


Location  Rainfall
0           Adelaide    4841.6
1             Albany    6827.8
2             Albury    5763.4
3       AliceSprings    2676.8
4      BadgerysCreek    6421.4
5           Ballarat    5268.8
6            Bendigo    4913.2
7           Brisbane    9941.0
8             Cairns   17157.2
9           Canberra    5953.2
10             Cobar    3368.4
11      CoffsHarbour   14946.6
12          Dartmoor    6315.2
13            Darwin   16260.2
14         GoldCoast   11232.8
15            Hobart    5106.6
16         Katherine    4993.7
17        Launceston    6092.3
18         Melbourne    4553.6
19  MelbourneAirport    4369.0
20           Mildura    2841.8
21             Moree    4652.6
22      MountGambier    6327.4
23       MountGinini    9570.6
24         Newcastle    9408.4
25              Nhil    1466.8
26         NorahHead    9921.4
27     NorfolkIsland    9270.4
28         Nuriootpa    4175.2
29        PearceRAAF    4610.0
30           Penrith    6447.6
31             Perth    6086.8
32      PerthAirport    5300.8
33          Portland    7581.0
34          Richmond    6310.6
35              Sale    4530.5
36        SalmonGums    3056.6
37            Sydney   11094.0
38     SydneyAirport    9044.8
39        Townsville   10571.8
40       Tuggeranong    6487.8
41             Uluru    1193.8
42        WaggaWagga    5088.8
43           Walpole    8194.4
44          Watsonia    5580.6
45       Williamtown    9168.1
46       Witchcliffe    8548.0
47        Wollongong   10720.0
48           Woomera    1466.8

In [68]:
print("After sorting: ")
total_rainfall_by_location.sort_values(by="Rainfall", ascending=True)

After sorting: 


Location  Rainfall
41             Uluru    1193.8
48           Woomera    1466.8
25              Nhil    1466.8
3       AliceSprings    2676.8
20           Mildura    2841.8
36        SalmonGums    3056.6
10             Cobar    3368.4
28         Nuriootpa    4175.2
19  MelbourneAirport    4369.0
35              Sale    4530.5
18         Melbourne    4553.6
29        PearceRAAF    4610.0
21             Moree    4652.6
0           Adelaide    4841.6
6            Bendigo    4913.2
16         Katherine    4993.7
42        WaggaWagga    5088.8
15            Hobart    5106.6
5           Ballarat    5268.8
32      PerthAirport    5300.8
44          Watsonia    5580.6
2             Albury    5763.4
9           Canberra    5953.2
31             Perth    6086.8
17        Launceston    6092.3
34          Richmond    6310.6
12          Dartmoor    6315.2
22      MountGambier    6327.4
4      BadgerysCreek    6421.4
30           Penrith    6447.6
40       Tuggeranong    6487.8
1             Albany    6827.8
33          Portland    7581.0
43           Walpole    8194.4
46       Witchcliffe    8548.0
38     SydneyAirport    9044.8
45       Williamtown    9168.1
27     NorfolkIsland    9270.4
24         Newcastle    9408.4
23       MountGinini    9570.6
26         NorahHead    9921.4
7           Brisbane    9941.0
39        Townsville   10571.8
47        Wollongong   10720.0
37            Sydney   11094.0
14         GoldCoast   11232.8
11      CoffsHarbour   14946.6
13            Darwin   16260.2
8             Cairns   17157.2

**Exercise**

Can you sort the `total_rainfall_by_location` dataframe by `Rainfall` in descending order?

In [69]:
# [TODO]
total_rainfall_by_location.sort_values(by="Rainfall", ascending=False)

Location  Rainfall
8             Cairns   17157.2
13            Darwin   16260.2
11      CoffsHarbour   14946.6
14         GoldCoast   11232.8
37            Sydney   11094.0
47        Wollongong   10720.0
39        Townsville   10571.8
7           Brisbane    9941.0
26         NorahHead    9921.4
23       MountGinini    9570.6
24         Newcastle    9408.4
27     NorfolkIsland    9270.4
45       Williamtown    9168.1
38     SydneyAirport    9044.8
46       Witchcliffe    8548.0
43           Walpole    8194.4
33          Portland    7581.0
1             Albany    6827.8
40       Tuggeranong    6487.8
30           Penrith    6447.6
4      BadgerysCreek    6421.4
22      MountGambier    6327.4
12          Dartmoor    6315.2
34          Richmond    6310.6
17        Launceston    6092.3
31             Perth    6086.8
9           Canberra    5953.2
2             Albury    5763.4
44          Watsonia    5580.6
32      PerthAirport    5300.8
5           Ballarat    5268.8
15            Hobart    5106.6
42        WaggaWagga    5088.8
16         Katherine    4993.7
6            Bendigo    4913.2
0           Adelaide    4841.6
21             Moree    4652.6
29        PearceRAAF    4610.0
18         Melbourne    4553.6
35              Sale    4530.5
19  MelbourneAirport    4369.0
28         Nuriootpa    4175.2
10             Cobar    3368.4
36        SalmonGums    3056.6
20           Mildura    2841.8
3       AliceSprings    2676.8
25              Nhil    1466.8
48           Woomera    1466.8
41             Uluru    1193.8

# Missing Values

As we have seen, there are lots of `NaN` values in the data. We have also learnt to use the `isnull()` and `notnull()` methods to check for missing values. 

**Dealing missing values** is something that data professionals do all the time. There are many ways to deal with missing values.
- Replace the missing values with some value (e.g. `mean`, `median` of the numerical columns or "Unknown" for categorical columns).
- Drop the rows with the missing values.

We'll learn about 2 very useful methods for dealing with missing values here: `fillna()` and `dropna()`.
- `fillna()`, as the name suggests`, will fill the missing values with a specified value.
- `dropna()`, as the name suggests`, will drop the rows with missing values.

Let's see what happen if we drop all rows with at least 1 missing value.

```python
weather_data_droppedna = weather_data.dropna()

print(f"Original shape of data: {weather_data.shape}")
print(f"New shape of data: {weather_data_droppedna.shape}")
```

In [70]:
weather_data_droppedna = weather_data.dropna()

print(f"Original shape of data: {weather_data.shape}")
print(f"New shape of data: {weather_data_droppedna.shape}")

Original shape of data: (145460, 28)
New shape of data: (56420, 28)


The data shrinks significantly (about 3 times) because we have dropped all rows with at least 1 missing value. This is generally **NOT A GOOD IDEA** because we lose too much valuable information. Normally, we only drop rows with missing values if we are very sure that many columns in the same rows are also missing.

**Exercise**

Can you tell me which column(s) have the most missing values?

In [71]:
# [TODO]
weather_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 28 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   Date                   145460 non-null  datetime64[ns]
 1   Location               145460 non-null  object        
 2   MinTemp                143975 non-null  float64       
 3   MaxTemp                144199 non-null  float64       
 4   Rainfall               142199 non-null  float64       
 5   Evaporation            82670 non-null   float64       
 6   Sunshine               75625 non-null   float64       
 7   WindGustDir            135134 non-null  object        
 8   WindGustSpeed          135197 non-null  float64       
 9   WindDir9am             134894 non-null  object        
 10  WindDir3pm             141232 non-null  object        
 11  WindSpeed9am           143693 non-null  float64       
 12  WindSpeed3pm           142398 non-null  floa

In [72]:
# [TODO]
weather_data.isna().sum(axis=0).sort_values(ascending=False)

Sunshine                 69835
Evaporation              62790
Cloud3pm                 59358
Cloud9am                 55888
Pressure9am              15065
Pressure3pm              15028
WindDir9am               10566
WindGustDir              10326
WindGustSpeed            10263
Humidity3pm               4507
WindDir3pm                4228
Temp3pm                   3609
RainTomorrow              3267
RainToday                 3261
NormalisedRainfall_v2     3261
NormalisedRainfall        3261
Rainfall                  3261
WindSpeed3pm              3062
Humidity9am               2654
WindSpeed9am              1767
Temp9am                   1767
MinTemp                   1485
MaxTemp                   1261
Month                        0
Date                         0
MinTemp > 5                  0
Location                     0
Year                         0
dtype: int64

**Exercise**

Let's choose column `Pressure9am` as an example and fill in the missing values using the `fillna()` method.

In order to decide which value to fill in, let's take a look at the summary statistics of that column.

Can you do that?

In [73]:
# [TODO]
weather_data["Pressure9am"].describe()

count    130395.00000
mean       1017.64994
std           7.10653
min         980.50000
25%        1012.90000
50%        1017.60000
75%        1022.40000
max        1041.00000
Name: Pressure9am, dtype: float64

We will choose to fill the missing values of `Pressure9am` with the `mean` of that column. Normally, I prefer to use `median` since `mean` is most often skewed by outliers. Nonetheless, in this case, the `mean` and `median` are very close to each other so we can use `mean` here.

We also choose to perform the operations `inplace` which means that we will modify the original dataframe.

```python
mean_Pressure9am = weather_data["Pressure9am"].mean()

weather_data.fillna({
    "Pressure9am": mean_Pressure9am,
}, inplace=True)
```

In [74]:
mean_Pressure9am = weather_data["Pressure9am"].mean()

weather_data.fillna({
    "Pressure9am": mean_Pressure9am,
}, inplace=True)

We can verify that the missing values have been filled in by running the following code:

```python
weather_data["Pressure9am"].isna().sum()
```

In [75]:
weather_data["Pressure9am"].isna().sum(axis=0)

0

# Combine Dataframes

It's very often that we retrieve data from multiple sources. For each source, we have data stored in a separate dataframe. Thus, it is crucial for us to know how to combine the dataframes.

`pandas` has 3 core methods to perform this task: `concat()`, `merge()` and `join()`. 
- The `concat()` method is used to combine multiple dataframes into one. 
- The `merge()` method is used to combine two dataframes based on a common column. 
- The `join()` method is used to combine two dataframes based on a common column.

Most of what `join()` can do is the same as `merge()`, so I'll only demonstrate the `merge()` method here.

The `join()` method combines two dataframes on the **basis of their indices** whereas the `merge()` method allows us to specify columns in addition to the indices to join on.
- You can read more about the `join()` method [here](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.join.html).
- You can read about the difference between `join()` and `merge()` [here](https://www.geeksforgeeks.org/what-is-the-difference-between-join-and-merge-in-pandas/).

There are various forms of join:
1. `inner`: Keep only elements that exist in both dataframes.
1. `outer`: Keep all elements from all dataframes.
1. `left`: Keep all elements from the left dataframe.
1. `right`: Keep all elements from the right dataframe.

Visual representations of various join types:

|Type of Join|Image|
|:-:|:-:|
|Inner|![image](../../images/inner_join.png)|
|Outer|![image](../../images/outer_join.png)|
|Left|![image](../../images/left_join.png)|
|Right|![image](../../images/right_join.png)|

Since we're familiar with the different types of join now, we will then see how to combine dataframes with `pandas`.

First, we'll create two dataframes:

```python
left_df = pd.DataFrame({
    "common_col": [1, 2, 3, 4, 5, 6, 7, 8],
    "left_value": [10, 20, 30, 40, 50, 60, 70, 80],
})

right_df = pd.DataFrame({
    "common_col": [3, 4, 5, 6, 9, 10, 11],
    "right_value": [-30, -40, -50, -60, -90, -100, -110],
})
```

In [76]:
left_df = pd.DataFrame({
    "common_col": [1, 2, 3, 4, 5, 6, 7, 8],
    "left_value": [10, 20, 30, 40, 50, 60, 70, 80],
})

right_df = pd.DataFrame({
    "common_col": [3, 4, 5, 6, 9, 10, 11],
    "right_value": [-30, -40, -50, -60, -90, -100, -110],
})

Let's take a look at the `left_df` and `right_df` dataframes.

In [78]:
left_df

common_col  left_value
0           1          10
1           2          20
2           3          30
3           4          40
4           5          50
5           6          60
6           7          70
7           8          80

In [79]:
right_df

common_col  right_value
0           3          -30
1           4          -40
2           5          -50
3           6          -60
4           9          -90
5          10         -100
6          11         -110

`inner` join `left_df` and `right_df` on the `common_col` column.

```python
inner_join_df = left_df.merge(right_df, on="common_col", how="inner")
```

In [80]:
inner_join_df = left_df.merge(right_df, on="common_col", how="inner")
inner_join_df

common_col  left_value  right_value
0           3          30          -30
1           4          40          -40
2           5          50          -50
3           6          60          -60

As expected, the `inner_join_df` dataframe contains **only common values that exist in both** `left_df` and `right_df` dataframes. 

`outer` join `left_df` and `right_df`.

```python
outer_join_df = left_df.merge(right_df, on="common_col", how="outer")
```

In [82]:
outer_join_df = left_df.merge(right_df, on="common_col", how="outer")
outer_join_df

common_col  left_value  right_value
0            1        10.0          NaN
1            2        20.0          NaN
2            3        30.0        -30.0
3            4        40.0        -40.0
4            5        50.0        -50.0
5            6        60.0        -60.0
6            7        70.0          NaN
7            8        80.0          NaN
8            9         NaN        -90.0
9           10         NaN       -100.0
10          11         NaN       -110.0

As expected, the `outer_join_df` dataframe contains **all common values that exist in either** `left_df` or `right_df` dataframes. 

`left` join `left_df` and `right_df`.

```python
left_join_df = left_df.merge(right_df, on="common_col", how="left")
```

In [83]:
left_join_df = left_df.merge(right_df, on="common_col", how="left")
left_join_df

common_col  left_value  right_value
0           1          10          NaN
1           2          20          NaN
2           3          30        -30.0
3           4          40        -40.0
4           5          50        -50.0
5           6          60        -60.0
6           7          70          NaN
7           8          80          NaN

As expected, the `left_join_df` contains all values from `left_df` and **only the common values that exist in both** `left_df` and `right_df` dataframes.

`right` join `left_df` and `right_df`.

```python
right_join_df = left_df.merge(right_df, on="common_col", how="right")
```

In [84]:
right_join_df = left_df.merge(right_df, on="common_col", how="right")
right_join_df

common_col  left_value  right_value
0           3        30.0          -30
1           4        40.0          -40
2           5        50.0          -50
3           6        60.0          -60
4           9         NaN          -90
5          10         NaN         -100
6          11         NaN         -110

Similarly, the `right_join_df` contains all values from `right_df` and **only the common values that exist in both** `left_df` and `right_df` dataframes.

**Concatenation** is a bit different from the merging techniques that you saw above. With concatenation, we are stitching dataframes together along an axis - either horizontally or vertically.

Visually, this is what concatenation looks like.

|Type of concat|Axis|Image|
|:-:|:-:|:-:|
|`verical`|`0` (`row axis`)|![image](../../images/concat_vertical_axis0.png)|
|`horizontal`|`1` (`column axis`)|![image](../../images/concat_horizontal_axis1.png)|

Let's see `concat()` in action!

In [85]:
vertical_concat = pd.concat([left_df, right_df], axis=0)
vertical_concat

common_col  left_value  right_value
0           1        10.0          NaN
1           2        20.0          NaN
2           3        30.0          NaN
3           4        40.0          NaN
4           5        50.0          NaN
5           6        60.0          NaN
6           7        70.0          NaN
7           8        80.0          NaN
0           3         NaN        -30.0
1           4         NaN        -40.0
2           5         NaN        -50.0
3           6         NaN        -60.0
4           9         NaN        -90.0
5          10         NaN       -100.0
6          11         NaN       -110.0

In [86]:
horizontal_concat = pd.concat([left_df, right_df], axis=1)
horizontal_concat

common_col  left_value  common_col  right_value
0           1          10         3.0        -30.0
1           2          20         4.0        -40.0
2           3          30         5.0        -50.0
3           4          40         6.0        -60.0
4           5          50         9.0        -90.0
5           6          60        10.0       -100.0
6           7          70        11.0       -110.0
7           8          80         NaN          NaN